src/mlpack/methods/reinforcement_learning/environment/maze.hpp

```diff
@@ -0,0 +1,406 @@
+            {
+                cellsInPath.push(lastCell);
+                cellsInPath.push(nextCell);
+                break;
+            }
+        }
+    }
+
+    // Pop last ({row,column}) from the stack
+      arma::ivec goalCell = cellsInPath.top();
+      cellsInPath.pop();
+
+    // Set the ({row,column}) asa goal cell
+      maze(goalCell(0),goalCell(1)) = 1;
+
+    // Set all other cells in stack to 0 
+    while (!cellsInPath.empty())
+    {
+        arma::ivec lastCell = cellsInPath.top();
+        cellsInPath.pop();
+      
+        maze(lastCell(0),lastCell(1)) = 0;
+    }
+
+    // Randomly set other cells to -1(wall) or 0(path)
+    for (size_t row = 0; row < rows; ++row)
+    {
+      for (size_t col = 0; col < columns; ++col)
+      {
+          if(maze(row,col) !=0 && maze(row,col) !=1 )
+          {
+            maze(row,col) = arma::randi(arma::distr_param(-1, 0));
+          }
+      }
+    }
+
+    // Store goal cell
+    goal = arma::vec({double(goalCell(0)),double(goalCell(1))});
+
+    // Store all the path cells as potential starting points
+    for (size_t row = 0; row < rows; ++row)
+    {
+      for (size_t col = 0; col < columns; ++col)
+      {
+          if(maze(row,col) == 0)
+          {
+            startingPoints.push_back(arma::vec({double(row),double(col)}));
+          }
+      }
+    }
+
+  }
+
+  /**
+   * Get reward and next state based on current
+   * state and current action.
+   *
+   * @param state The current state.
+   * @param action The current action.
+   * @param nextState The next state.
+   * @param transitionGoal The goal for transition
+   * @return reward, it's always 1.0.
+   */
+  double Sample(const State& state,
+                const Action& action,
+                State& nextState,
+                const State& transitionGoal)
+  {
+    // Update the number of steps performed.
+    stepsPerformed++;
+
+    // Make a vector to estimate nextstate.
+    arma::vec currentState {state.Row(), state.Column()};
+    arma::vec direction = directions[action.action];
+
+    arma::vec currentNextState = currentState + direction;
+    nextState.Row() = currentNextState[0];
+    nextState.Column() = currentNextState[1];
+
+    // dont move to that position if invalid
+    bool invalid = false;
+
+    if(currentNextState(0)>= maze.n_rows || currentNextState(1)>= maze.n_cols
+        || currentNextState(0) < 0 || currentNextState(1) < 0 || 
+        maze(currentNextState(0),currentNextState(1))==-1 ){
+        invalid = true;
+        nextState.Row() = currentState[0];
+        nextState.Column() = currentState[1];
+    }   
+
+    // Check if the episode has terminated.
+    bool done = IsTerminal(nextState);
+
+    // Do not reward agent if it failed.
+    if (done && maxSteps != 0 && stepsPerformed >= maxSteps)
+      return 0.0;
+    
+    if (done && maze(nextState.Row(),nextState.Column()) == 1 && 
+        nextState.Row() == goal(0) && nextState.Column() == goal(1) )
```

Comment:
nextState == goal?

Reply:
Its going to give me a vector rather than a bool

Reply:
remember `arma::approx_equal`?

Reply:
Yeah I do. I just thought that is pretty straight forward. I will make that change then ?

Reply:
K